### ISSUE TITLE
Feature: find cuesheets for songs I have

### ISSUE THREAD

**Issue by mpogue2** [mpogue2/SquareDesk#225]
See also #55 for a discussion about this.  We have downloadable ZIP file cuesheets now, and that's great.  But, it could be even better!  We could have a menu item that when selected will fetch matching cuesheets (all of them) over the internet, just downloading the matches (not all of them).

This will minimize the amount of cuesheets on disk, saving a lot of space.  It's probably cheap to host, too, something like Amazon Lambda could probably do it for almost no cost.  

Pass up a JSON list of songs and the cuesheets that we already have, and it returns a ZIP-compressed archive of cuesheets that we don't have already, that can be unpacked directly into the lyrics directory (overlaying what's there already, since we know there are no cuesheet conflicts).

Maybe: Lyrics > Get matching lyrics..., which brings up a dialog with a progress bar.  Something like that.

As a bonus, we could do the PDF/DOC/DOCX translation to HTML (automated, once), and just fetch those HTML files.  It would save a bunch of time to cut/copy/paste from formats we don't yet support in the app itself!  We could also auto-expand the lyrics into a scrollable format, which could either replace the regular format, or be beside the regular format.

**Comment by mpogue2** [comment]
We could also provide a separate Lambda interface for music providers, who can upload/replace cuesheets for their music.  It becomes a community-maintained database of cuesheets.

**Comment by mpogue2** [comment]
On further reflection, there are two fundamentally different approaches we could take here:

1) The app sends a compressed list of filenames/cuesheets that it has to the Cloud (probably ~10KB), and the Cloud returns a ZIP file containing all the matching cuesheets that the app doesn't have
ADV: one place for all the files (Cloud), and they can live there compressed.
DIS: matcher on Cloud might not be the same as matcher in the app, app changes faster.
DIS: uploads speeds (~10K upload here) are typically slower than download speeds
DIS: Cloud has smarts, can be AWS Lambda, but it's dynamic content

2) The app fetches a compressed list of filenames that the Cloud has (probably ~100KB), the APP matches them up, and then makes individual (or later, bulk) requests for the compressed cuesheets that it doesn't have.
ADV: one place for the matcher (in the app), can be updated at any time without Cloud changes
DIS: larger initial request size (~100K compressed downloaded from Cloud)
ADV: download speeds are much faster than upload speeds anyway
ADV: Cloud becomes just an http server of static content
DIS: Files need to live there uncompressed, so they can be fetched one at a time (e.g. squaredesk.net/cuesheets/foo.htm), unless we want to write a CGI-BIN to extract them, or we could host the extractor on AWS Lambda (OTOH, the total size of all cuesheets is only about 35MB uncompressed).

I'm leaning to option 2, uncompressed static files, hosted at http://squaredesk.net, with a compressed index file that can be fetched in a single http request.  There will also be a file
containing a checksum or last update date time of the index file, so that the index file is only fetched when the checksum changes (to reduce download traffic).  We can get fancier later.

**Comment by danlyke** [comment]
Most web technologies are pretty good about dealing with gzipped files, so you could always do them as individual .gz static files. This still leaves them to be fetched on at a time, but minimizes server processing load.

**Comment by mpogue2** [comment]
AH, I completely forgot about that option...  I can just turn on the built-in gzip compression, and everything will be MUCH smaller.  In fact, I've been running all my websites uncompressed forever (doh!).  Turning on compression, I get between 40-60% savings in bandwidth, just by one click in the control panel.  Next, I'll see how compressible the individual cuesheets are....

OK, now I'm leaning to Option 3: uncompressed static files hosted at http://squaredesk.net that are dynamically compressed by mod_gzip, with either a) a compressed index file that can be fetched in a single http request, or b) I just allow directory listing for the directory containing the files, and fetch that once (maybe it will be automatically compressed, too?  I will check.).

More testing coming.

**Comment by mpogue2** [comment]
Fetching a small list of files (using the normal Apache directory listing service) is compressed automatically by about 64% (1K -> 0.4K, for a listing of 5 files).

A typical HTML file is automatically compressed by about 70% (1K -> 3.5K).
A typical PDF file is automatically compressed by about 5% (190K -> 180K).
A typical DOCX file is automatically compressed by about 20% (13.3K -> 10.6K).

I'm not worried about CPU time to compress (and I think mod_gzip caches the results anyway, right?), I'm more worried about bandwidth and latency to download.

This looks good to me, so I'm going to start with Option 3b: uncompressed static files hosted at http://squaredesk.net that are automatically compressed by mod_gzip, with an automatically-compressed directory listing for the directory containing the files.  The directory listing returns HTML, but I think it's straightforward to parse.

Thanks, @danlyke for the gzip suggestion!

```html
<html>
 <head>
  <title>Index of /cuesheets</title>
 </head>
 <body>
  <h1>Index of /cuesheets</h1>
    <ul>
    <li><a href="/"> Parent Directory</a></li>
    <li><a href="EGO%20125%20-%20Jump%20In%20The%20Line%20-%20Shake%20Senora%20(Cuesheet).html"> EGO 125 - Jump In The Line - Shake Senora (Cuesheet).html</a></li>
etc...
    </ul>
 </body>
</html>
```

**Comment by mpogue2** [comment]
Implemented by c965d524476315aff71e9546d90e73248ea5a88b on Master.

Cuesheets are hosted at http://squaredesk.net/cuesheets now.  First, a (compressed) directory listing is downloaded and expanded.  Second, we do fuzzy matching (same as today's fuzzy matching) of songs vs cuesheets.  Those cuesheets that match are downloaded (in compressed form) and expanded to "<musicDirectory>/lyrics/downloaded", if they are not there already.  Finally, we rescan the <musicDirectory> for files, and reload the lyrics dropdown (because we might have more cuesheets that match now).

This should also speed up both startup loading and double-click loading, since there are many fewer cuesheet files needed now.
